File: AwesomeSpotlightViewDemoObjC/AwesomeSpotlightViewDemoObjC/ViewController.m
//
//  ViewController.m
//  AwesomeSpotlightViewDemoObjC
//
//  Created by FromF on 2018/08/15.
//  Copyright © 2018年 FromF. All rights reserved.
//

#import "ViewController.h"
#import "AwesomeSpotlightViewDemoObjC-Swift.h"

@interface ViewController () <AwesomeSpotlightViewDelegate>
@property (weak, nonatomic) IBOutlet UIImageView *logoImageView;
@property (weak, nonatomic) IBOutlet UILabel *nameLabel;

@property (weak, nonatomic) IBOutlet UIButton *showButton;
@property (weak, nonatomic) IBOutlet UIButton *showWithContinueAndSkipButton;
@property (weak, nonatomic) IBOutlet UIButton *showAllAtOnceButton;

@property (strong, nonatomic) AwesomeSpotlightView *spotlightView;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setupViews];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    [self setupSpotlight];
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

- (void)setupViews {
    self.showButton.layer.cornerRadius = 8.0;
    self.showButton.clipsToBounds = true;
    
    self.showWithContinueAndSkipButton.layer.cornerRadius = 8.0;
    self.showWithContinueAndSkipButton.clipsToBounds = true;
    
    self.showAllAtOnceButton.layer.cornerRadius = 8.0;
    self.showAllAtOnceButton.clipsToBounds = true;
}

- (void)setupSpotlight {
    UIEdgeInsets spotlightMarginDefault = UIEdgeInsetsMake(0.0, 0.0, 0.0, 0.0);
    
    CGRect logoImageViewSpotlightRect = CGRectMake(self.logoImageView.frame.origin.x, self.logoImageView.frame.origin.y, self.logoImageView.frame.size.width, self.logoImageView.frame.size.height);
    UIEdgeInsets logoImageViewSpotlightMargin = UIEdgeInsetsMake(20.0, 20.0, 20.0, 20.0);
    AwesomeSpotlight *logoImageViewSpotlight = [[AwesomeSpotlight alloc] initWithRect:logoImageViewSpotlightRect shape:AwesomeSpotlightShapeCircle text:@"logoImageViewSpotlight" margin:logoImageViewSpotlightMargin isAllowPassTouchesThroughSpotlight:false];
    
    AwesomeSpotlight *nameLabelSpotlight = [[AwesomeSpotlight alloc] initWithRect:self.nameLabel.frame shape:AwesomeSpotlightShapeRectangle text:@"nameLabelSpotlight" margin:spotlightMarginDefault isAllowPassTouchesThroughSpotlight:false];
    
    AwesomeSpotlight *showButtonSpotSpotlight = [[AwesomeSpotlight alloc] initWithRect:self.showButton.frame shape:AwesomeSpotlightShapeRoundRectangle text:@"showButtonSpotSpotlight" margin:spotlightMarginDefault isAllowPassTouchesThroughSpotlight:false];

    AwesomeSpotlight *showWithContinueAndSkipButtonSpotlight = [[AwesomeSpotlight alloc] initWithRect:self.showWithContinueAndSkipButton.frame shape:AwesomeSpotlightShapeRoundRectangle text:@"showWithContinueAndSkipButtonSpotlight" margin:spotlightMarginDefault isAllowPassTouchesThroughSpotlight:false];
    
    AwesomeSpotlight *showAllAtOnceButtonSpotlight = [[AwesomeSpotlight alloc] initWithRect:self.showAllAtOnceButton.frame shape:AwesomeSpotlightShapeRoundRectangle text:@"showAllAtOnceButtonSpotlight" margin:spotlightMarginDefault isAllowPassTouchesThroughSpotlight:false];
    
    self.spotlightView = [[AwesomeSpotlightView alloc] initWithFrame:self.view.frame spotlight:@[logoImageViewSpotlight, nameLabelSpotlight, showButtonSpotSpotlight, showWithContinueAndSkipButtonSpotlight, showAllAtOnceButtonSpotlight]];
    self.spotlightView.cutoutRadius = 8;
    self.spotlightView.delegate = self;
}

#pragma mark Actions
- (IBAction)handleShowAction:(id)sender {
    [self.view addSubview:self.spotlightView];
    [self.spotlightView setContinueButtonEnable:false];
    [self.spotlightView setSkipButtonEnable:false];
    self.spotlightView.showAllSpotlightsAtOnce = false;
    [self.spotlightView start];
}

- (IBAction)handleShowWithContinueAndSkipAction:(id)sender {
    [self.view addSubview:self.spotlightView];
    [self.spotlightView setContinueButtonEnable:true];
    [self.spotlightView setSkipButtonEnable:true];
    self.spotlightView.showAllSpotlightsAtOnce = false;
    [self.spotlightView start];
}

- (IBAction)handleShowAllAtOnceAction:(id)sender {
    [self.view addSubview:self.spotlightView];
    [self.spotlightView setContinueButtonEnable:false];
    [self.spotlightView setSkipButtonEnable:false];
    self.spotlightView.showAllSpotlightsAtOnce = true;
    [self.spotlightView start];
}

#pragma mark AwesomeSpotlightViewDelegate
- (void)spotlightView:(AwesomeSpotlightView *)spotlightView willNavigateToIndex:(NSInteger)index {
    NSLog(@"spotlightView willNavigateToIndex index =  %ld",(long)index);
}

- (void)spotlightView:(AwesomeSpotlightView *)spotlightView didNavigateToIndex:(NSInteger)index {
    NSLog(@"spotlightView didNavigateToIndex index =  %ld",(long)index);
}

- (void)spotlightViewWillCleanup:(AwesomeSpotlightView *)spotlightView atIndex:(NSInteger)index {
    NSLog(@"spotlightViewWillCleanup atIndex =  %ld",(long)index);
}

- (void)spotlightViewDidCleanup:(AwesomeSpotlightView *)spotlightView {
    NSLog(@"spotlightViewDidCleanup");
}

@end

File: AwesomeSpotlightViewDemo/AwesomeSpotlightViewDemo/ViewController.swift
//
//  ViewController.swift
//  AwesomeSpotlightView
//
//  Created by Alex Shoshiashvili on 24.02.17.
//  Copyright © 2017 Alex Shoshiashvili. All rights reserved.
//

import UIKit
import AwesomeSpotlightView

class ViewController: UIViewController {
  
  @IBOutlet weak var logoImageView: UIImageView!
  @IBOutlet weak var nameLabel: UILabel!
  
  @IBOutlet weak var showButton: UIButton!
  @IBOutlet weak var showWithContinueAndSkipButton: UIButton!
  @IBOutlet weak var showAllAtOnceButton: UIButton!
  
  var spotlightView = AwesomeSpotlightView()
  
  override func viewDidLoad() {
    super.viewDidLoad()
    setupViews()
  }
  
  override func viewDidAppear(_ animated: Bool) {
    super.viewDidAppear(animated)
    setupSpotlight()
  }
  
  override func didReceiveMemoryWarning() {
    super.didReceiveMemoryWarning()
  }
  
  func setupViews() {
    showButton.layer.cornerRadius = 8.0
    showButton.clipsToBounds = true
    
    showWithContinueAndSkipButton.layer.cornerRadius = 8.0
    showWithContinueAndSkipButton.clipsToBounds = true
    
    showAllAtOnceButton.layer.cornerRadius = 8.0
    showAllAtOnceButton.clipsToBounds = true
  }
  
  func setupSpotlight() {
    let logoImageViewSpotlightRect = CGRect(x: logoImageView.frame.origin.x, y: logoImageView.frame.origin.y, width: logoImageView.frame.size.width, height: logoImageView.frame.size.height)
    let logoImageViewSpotlightMargin = UIEdgeInsets(top: 20, left: 20, bottom: 20, right: 20)
    let logoImageViewSpotlight = AwesomeSpotlight(withRect: logoImageViewSpotlightRect, shape: .circle, text: "logoImageViewSpotlight", margin: logoImageViewSpotlightMargin)
    
    let nameLabelSpotlight = AwesomeSpotlight(withRect: nameLabel.frame, shape: .rectangle, text: "nameLabelSpotlight")
    
    let showButtonSpotSpotlight = AwesomeSpotlight(withRect: showButton.frame, shape: .roundRectangle, text: "showButtonSpotSpotlight")
    
    let showWithContinueAndSkipButtonSpotlight = AwesomeSpotlight(withRect: showWithContinueAndSkipButton.frame, shape: .roundRectangle, text: "showWithContinueAndSkipButtonSpotlight")
    
    let showAllAtOnceButtonSpotlight = AwesomeSpotlight(withRect: showAllAtOnceButton.frame, shape: .roundRectangle, text: "showAllAtOnceButtonSpotlight", isAllowPassTouchesThroughSpotlight: true)
    
    spotlightView = AwesomeSpotlightView(frame: view.frame, spotlight: [logoImageViewSpotlight, nameLabelSpotlight, showButtonSpotSpotlight, showWithContinueAndSkipButtonSpotlight, showAllAtOnceButtonSpotlight])
    spotlightView.cutoutRadius = 8
    spotlightView.delegate = self
  }
  
  // MARK: - Actions
  
  @IBAction func handleShowAction(_ sender: AnyObject) {
    view.addSubview(spotlightView)
    spotlightView.continueButtonModel.isEnable = false
    spotlightView.skipButtonModel.isEnable = false
    spotlightView.showAllSpotlightsAtOnce = false
    spotlightView.start()
  }
  @IBAction func handleShowWithContinueAndSkipAction(_ sender: AnyObject) {
    view.addSubview(spotlightView)
    spotlightView.continueButtonModel.isEnable = true
    spotlightView.skipButtonModel.isEnable = true
    spotlightView.showAllSpotlightsAtOnce = false
    
    // Uncomment if want to customize skip button
    //spotlightView.skipButtonLastStepTitle = "Finish"
    //spotlightView.skipButtonModel = AwesomeTabButton(title: "Skip".uppercased(),
    //                                                 font: UIFont.boldSystemFont(ofSize: 16.0),
    //                                                 isEnable: true,
    //                                                 backgroundColor: .red)
    
    // Uncomment if want to customize continue button
    //spotlightView.continueButtonModel = AwesomeTabButton(title: "Next",
    //                                                     font: UIFont.italicSystemFont(ofSize: 16.0),
    //                                                     isEnable: true,
    //                                                     backgroundColor: .blue)
    
    spotlightView.start()
  }
  @IBAction func handleShowAllAtOnceAction(_ sender: AnyObject) {
    view.addSubview(spotlightView)
    spotlightView.continueButtonModel.isEnable = false
    spotlightView.skipButtonModel.isEnable = false
    spotlightView.showAllSpotlightsAtOnce = true
    spotlightView.start()
  }
  
}

extension ViewController : AwesomeSpotlightViewDelegate {
  
  func spotlightView(_ spotlightView: AwesomeSpotlightView, willNavigateToIndex index: Int) {
    print("spotlightView willNavigateToIndex index = \(index)")
  }
  
  func spotlightView(_ spotlightView: AwesomeSpotlightView, didNavigateToIndex index: Int) {
    print("spotlightView didNavigateToIndex index = \(index)")
  }
  
  func spotlightViewWillCleanup(_ spotlightView: AwesomeSpotlightView, atIndex index: Int) {
    print("spotlightViewWillCleanup atIndex = \(index)")
  }
  
  func spotlightViewDidCleanup(_ spotlightView: AwesomeSpotlightView) {
    print("spotlightViewDidCleanup")
  }
  
}
